### PR TITLE
TRUNK-6008: Fix HibernatePatientDAO.getDuplicatePatientsByAttributes

### PR DESCRIPTION
**Pull request by BartlomiejRasztabiga**
## Description of what I changed
Fixed issues with SQLException while trying to get patients with duplicated names. I've also added tests for this functionality.

## Issue I worked on
see https://issues.openmrs.org/browse/TRUNK-6008

## Checklist: I completed these to help reviewers :)
- [X] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [X] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [X] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [X] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [X] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 1]
```diff
@@ -478,8 +478,8 @@ private String getDuplicatePatientsSQLString(List<String> attributes) {
 				innerFields.add("person1." + attribute);
 			} else if (personNameFieldNames.contains(attribute)) {
 				if (!outerSelect.toString().contains("person_name")) {
-					outerSelect.append("inner join person_name t3 on t2.person_id = t3.person_id ");
-					innerSelect.append("inner join person_name pn1 on person1.person_id = pn1.person_id ");
+					outerSelect.append("inner join person_name t3 on t1.patient_id = t3.person_id ");
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:481]: Do you mind explaining this change?

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java [hunk 2]
```diff
@@ -63,4 +70,116 @@ public void getPatientIdentifiers_shouldGetByPatients() {
...
+		// when
+		List<Patient> duplicatePatients = hibernatePatientDao
+			.getDuplicatePatientsByAttributes(Arrays.asList("gender", "birthdate", "givenName", "middleName", "familyName"));
+
+		// then
+		assertThat(duplicatePatients.size(), equalTo(2));
+	}
+
+	@Test
+	public void getDuplicatePatientsByAttributes_shouldReturnPatientsWithDuplicatedAllButOneFields() {
```
Comment (HerbertYiga) [inline api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java:162]: @BartlomiejRasztabiga  could this need a better wording, failing t o understand it

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java (hunk 1 above)
Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:481]: Sure. Before that's how the SQL string looked like:
`select distinct t1.patient_id from patient t1 inner join person_name t3 on t2.person_id = t3.person_id , (Select pn1.family_name from patient p1 inner join person_name pn1 on person1.person_id = pn1.person_id  where p1.voided = false  group by pn1.family_name having count(*) > 1 order by pn1.family_name) t5 where  t3.family_name = t5.family_name;`

It had two syntax errors:
- first, there is no t2 alias defined anywhere,  I've changed that to t1 which is an alias for patient table. It also doesn't contain person_id column - now it's called patient_id
- second, there is no person2 alias defined, I've changed that to p1 which is an alias used in this inner join. It also doesn't contain person_id column - now it's called patient_id

That's how the SQL string looks now:
`select distinct t1.patient_id from patient t1 inner join person_name t3 on t1.patient_id = t3.person_id , (Select pn1.family_name from patient p1 inner join person_name pn1 on p1.patient_id = pn1.person_id  where p1.voided = false  group by pn1.family_name having count(*) > 1 order by pn1.family_name) t5 where  t3.family_name = t5.family_name;`

Basically, the initial thought behind this was to inner join `person_name` table with `patient` table. The previous code used wrong aliases and column names so I fixed those.

File: api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java (hunk 2 above)
Comment (BartlomiejRasztabiga) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java:162]: Sure I will find a better one. I wanted it to mean that those two patients have all duplicated fields besides gender :)

Comment (BartlomiejRasztabiga) [reply api/src/test/java/org/openmrs/api/db/hibernate/HibernatePatientDAOTest.java]: I've updated this test's name in the latest commit

File: api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java [hunk 3]
```diff
@@ -478,8 +478,8 @@ private String getDuplicatePatientsSQLString(List<String> attributes) {
 				innerFields.add("person1." + attribute);
 			} else if (personNameFieldNames.contains(attribute)) {
 				if (!outerSelect.toString().contains("person_name")) {
-					outerSelect.append("inner join person_name t3 on t2.person_id = t3.person_id ");
```
Comment (dkayiwa) [inline api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:481]: Will all the newly introduced tests fail without this change?

Comment (BartlomiejRasztabiga) [reply api/src/main/java/org/openmrs/api/db/hibernate/HibernatePatientDAO.java:481]: Not every one of them fails, there are 2 that pass. These are the tests that use "gender" attribute first, so then the `t2` alias is already present in the resulting SQL. However, I think my solution is correct. You can look at other branches of this `if-else` statement, as they all look similar to mine.
